File: src/engine.ts
// Reflex — Execution Engine
// Implements DESIGN.md Section 3.2
// M4-1: Constructor, init(), state inspection.
// M4-2: step() — single-workflow stepping with event emission.
// run() stubbed — implemented in M4-4.

import {
  Workflow,
  Node,
  Edge,
  BlackboardReader,
  BlackboardSource,
  StackFrame,
  DecisionAgent,
  DecisionContext,
  Decision,
  StepResult,
  RunResult,
  EngineEvent,
  EventHandler,
  EngineStatus,
} from './types';
import { WorkflowRegistry } from './registry';
import { ScopedBlackboard, ScopedBlackboardReader } from './blackboard';
import { filterEdges } from './guards';

// ---------------------------------------------------------------------------
// Engine Error
// ---------------------------------------------------------------------------

export class EngineError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EngineError';
  }
}

// ---------------------------------------------------------------------------
// ReflexEngine
// ---------------------------------------------------------------------------

export class ReflexEngine {
  private readonly _registry: WorkflowRegistry;
  private readonly _agent: DecisionAgent;

  // Session state — null until init() is called
  private _sessionId: string | null = null;
  private _status: EngineStatus = 'idle';
  private _currentWorkflowId: string | null = null;
  private _currentNodeId: string | null = null;
  private _currentBlackboard: ScopedBlackboard | null = null;

  // Call stack — suspended workflow frames (active frame is NOT on the stack)
  private _stack: StackFrame[] = [];

  // Event handlers
  private readonly _handlers: Map<EngineEvent, EventHandler[]> = new Map();

  constructor(registry: WorkflowRegistry, agent: DecisionAgent) {
    this._registry = registry;
    this._agent = agent;
  }

  // -------------------------------------------------------------------------
  // Lifecycle
  // -------------------------------------------------------------------------

  async init(workflowId: string): Promise<string> {
    const workflow = this._registry.get(workflowId);
    if (!workflow) {
      throw new EngineError(
        `Cannot initialize: workflow '${workflowId}' is not registered`,
      );
    }

    this._sessionId = crypto.randomUUID();
    this._currentWorkflowId = workflowId;
    this._currentNodeId = workflow.entry;
    this._currentBlackboard = new ScopedBlackboard();
    this._stack = [];
    this._status = 'running';

    return this._sessionId;
  }

  async step(): Promise<StepResult> {
    // -- Precondition guards ------------------------------------------------
    if (this._status !== 'running' && this._status !== 'suspended') {
      throw new EngineError(
        `step() called in invalid state: '${this._status}' — engine must be 'running' or 'suspended'`,
      );
    }
    if (
      this._currentWorkflowId === null ||
      this._currentNodeId === null ||
      this._currentBlackboard === null
    ) {
      throw new EngineError('step() called before init()');
    }
    // Resume from suspension
    if (this._status === 'suspended') {
      this._status = 'running';
    }

    const workflow = this._registry.get(this._currentWorkflowId)!;
    const node = workflow.nodes[this._currentNodeId]!;

    // -- Guard evaluation ---------------------------------------------------
    const reader = this.blackboard();
    const filterResult = filterEdges(
      this._currentNodeId,
      workflow.edges,
      reader,
    );
    if (!filterResult.ok) {
      this._status = 'suspended';
      this._emit('engine:error', {
        error: filterResult.error,
        nodeId: this._currentNodeId,
      });
      return { status: 'suspended', reason: 'Guard evaluation error' };
    }
    const validEdges = filterResult.edges;

    // -- Build DecisionContext and call agent --------------------------------
    const context: DecisionContext = {
      workflow,
      node,
      blackboard: reader,
      validEdges,
      stack: this.stack(),
    };

    let decision: Decision;
    try {
      decision = await this._agent.resolve(context);
    } catch (error) {
      this._status = 'suspended';
      this._emit('engine:error', { error, nodeId: this._currentNodeId });
      return { status: 'suspended', reason: 'Decision agent threw an error' };
    }

    // -- Handle advance -----------------------------------------------------
    if (decision.type === 'advance') {
      const chosenEdge = validEdges.find((e) => e.id === decision.edge);
      if (!chosenEdge) {
        this._status = 'suspended';
        this._emit('engine:error', {
          error: new EngineError(
            `Decision agent chose invalid edge '${decision.edge}'`,
          ),
          nodeId: this._currentNodeId,
        });
        return { status: 'suspended', reason: 'Invalid edge selection' };
      }

      this._emit('node:exit', { node, workflow });
      this._emit('edge:traverse', { edge: chosenEdge, workflow });

      if (decision.writes && decision.writes.length > 0) {
        const source: BlackboardSource = {
          workflowId: this._currentWorkflowId,
          nodeId: this._currentNodeId,
          stackDepth: this._stack.length,
        };
        const newEntries = this._currentBlackboard.append(
          decision.writes,
          source,
        );
        this._emit('blackboard:write', { entries: newEntries, workflow });
      }

      this._currentNodeId = chosenEdge.to;
      const nextNode = workflow.nodes[chosenEdge.to]!;
      this._emit('node:enter', { node: nextNode, workflow });

      return { status: 'advanced', node: nextNode };
    }

    // -- Handle suspend -----------------------------------------------------
    if (decision.type === 'suspend') {
      this._status = 'suspended';
      this._emit('engine:suspend', {
        reason: decision.reason,
        nodeId: this._currentNodeId,
      });
      return { status: 'suspended', reason: decision.reason };
    }

    // -- Handle complete ----------------------------------------------------
    // Enforce terminal-node-only (structural: no outgoing edges)
    const hasOutgoing = workflow.edges.some(
      (e) => e.from === this._currentNodeId,
    );
    if (hasOutgoing) {
      this._status = 'suspended';
      this._emit('engine:error', {
        error: new EngineError(
          `Decision agent returned 'complete' at non-terminal node '${this._currentNodeId}'`,
        ),
        nodeId: this._currentNodeId,
      });
      return { status: 'suspended', reason: 'complete at non-terminal node' };
    }

    if (decision.writes && decision.writes.length > 0) {
      const source: BlackboardSource = {
        workflowId: this._currentWorkflowId,
        nodeId: this._currentNodeId,
        stackDepth: this._stack.length,
      };
      const newEntries = this._currentBlackboard.append(
        decision.writes,
        source,
      );
      this._emit('blackboard:write', { entries: newEntries, workflow });
    }

    if (this._stack.length === 0) {
      this._status = 'completed';
      this._emit('engine:complete', { workflow });
      return { status: 'completed' };
    }

    // Stack pop — M4-3 implements this branch
    throw new EngineError(
      'complete with non-empty stack not implemented — see M4-3',
    );
  }

  async run(): Promise<RunResult> {
    throw new EngineError('run() not implemented — see M4-4');
  }

  // -------------------------------------------------------------------------
  // State Inspection
  // -------------------------------------------------------------------------

  status(): EngineStatus {
    return this._status;
  }

  sessionId(): string {
    if (this._sessionId === null) {
      throw new EngineError('Engine not initialized — call init() first');
    }
    return this._sessionId;
  }

  currentNode(): Node | null {
    if (this._currentWorkflowId === null || this._currentNodeId === null) {
      return null;
    }
    const workflow = this._registry.get(this._currentWorkflowId);
    if (!workflow) return null;
    return workflow.nodes[this._currentNodeId] ?? null;
  }

  currentWorkflow(): Workflow | null {
    if (this._currentWorkflowId === null) return null;
    return this._registry.get(this._currentWorkflowId) ?? null;
  }

  blackboard(): BlackboardReader {
    if (this._currentBlackboard === null) {
      return new ScopedBlackboardReader([]);
    }
    // Stack frames ordered so _stack[0] is the most-recent parent.
    // ScopedBlackboard.reader() takes parent scopes in that same order.
    const parentScopes = this._stack.map((frame) => [...frame.blackboard]);
    return this._currentBlackboard.reader(parentScopes);
  }

  stack(): ReadonlyArray<StackFrame> {
    return [...this._stack];
  }

  validEdges(): Edge[] {
    const workflow = this.currentWorkflow();
    if (!workflow || this._currentNodeId === null) return [];

    const reader = this.blackboard();
    const result = filterEdges(this._currentNodeId, workflow.edges, reader);

    if (!result.ok) {
      // Guard evaluation error — proper engine:error event emission is
      // implemented in M4-5. For now, return no valid edges.
      return [];
    }

    return result.edges;
  }

  // -------------------------------------------------------------------------
  // Events
  // -------------------------------------------------------------------------

  on(event: EngineEvent, handler: EventHandler): void {
    const handlers = this._handlers.get(event) ?? [];
    handlers.push(handler);
    this._handlers.set(event, handlers);
  }

  private _emit(event: EngineEvent, payload?: unknown): void {
    const handlers = this._handlers.get(event);
    if (handlers) {
      for (const handler of handlers) {
        handler(payload);
      }
    }
  }
}

File: go/engine.go
package reflex

import (
	"context"
	"crypto/rand"
	"fmt"
	"encoding/hex"
)

// EngineError represents an error from the execution engine.
type EngineError struct {
	Message string
}

func (e *EngineError) Error() string { return e.Message }

// Engine is the Reflex execution engine. It steps through workflow DAGs,
// manages the call stack for sub-workflow composition, and emits events.
//
// See DESIGN.md Section 3.2.
type Engine struct {
	registry *Registry
	agent    DecisionAgent

	sessionID        string
	status           EngineStatus
	currentWorkflowID string
	currentNodeID    string
	currentBlackboard *ScopedBlackboard
	stack            []StackFrame
	skipInvocation   bool

	handlers map[EventType][]EventHandler
}

// NewEngine creates an engine bound to a registry and decision agent.
func NewEngine(registry *Registry, agent DecisionAgent) *Engine {
	return &Engine{
		registry: registry,
		agent:    agent,
		status:   StatusIdle,
		handlers: make(map[EventType][]EventHandler),
	}
}

// On registers an event handler for the given event type.
func (e *Engine) On(event EventType, handler EventHandler) {
	e.handlers[event] = append(e.handlers[event], handler)
}

// Init initializes a new session for the given workflow.
// Returns the session ID.
func (e *Engine) Init(workflowID string) (string, error) {
	w, ok := e.registry.Get(workflowID)
	if !ok {
		return "", &EngineError{Message: fmt.Sprintf("cannot initialize: workflow '%s' is not registered", workflowID)}
	}

	e.sessionID = generateUUID()
	e.currentWorkflowID = workflowID
	e.currentNodeID = w.Entry
	e.currentBlackboard = NewBlackboard()
	e.stack = nil
	e.skipInvocation = false
	e.status = StatusRunning

	return e.sessionID, nil
}

// Step executes one iteration of the execution loop.
func (e *Engine) Step(ctx context.Context) (StepResult, error) {
	// Precondition guards
	if e.status != StatusRunning && e.status != StatusSuspended {
		return StepResult{}, &EngineError{
			Message: fmt.Sprintf("step() called in invalid state: '%s'", e.status),
		}
	}
	if e.currentBlackboard == nil {
		return StepResult{}, &EngineError{Message: "step() called before init()"}
	}
	if e.status == StatusSuspended {
		e.status = StatusRunning
	}

	w, _ := e.registry.Get(e.currentWorkflowID)
	node := w.Nodes[e.currentNodeID]

	// -- Invocation node handling --
	if node.Invokes != nil && !e.skipInvocation {
		subW, ok := e.registry.Get(node.Invokes.WorkflowID)
		if !ok {
			e.status = StatusSuspended
			e.emit(EventEngineError, Event{
				Type:   EventEngineError,
				NodeID: e.currentNodeID,
				Reason: fmt.Sprintf("sub-workflow '%s' not found", node.Invokes.WorkflowID),
			})
			return StepResult{
				Status: StepSuspended,
				Reason: fmt.Sprintf("sub-workflow '%s' not found", node.Invokes.WorkflowID),
			}, nil
		}

		// Push current frame
		frame := StackFrame{
			WorkflowID:    e.currentWorkflowID,
			CurrentNodeID: e.currentNodeID,
			ReturnMap:     node.Invokes.ReturnMap,
			Blackboard:    e.currentBlackboard.Entries(),
		}
		e.stack = append([]StackFrame{frame}, e.stack...)

		// Start sub-workflow
		e.currentWorkflowID = subW.ID
		e.currentNodeID = subW.Entry
		e.currentBlackboard = NewBlackboard()

		e.emit(EventWorkflowPush, Event{Type: EventWorkflowPush, WorkflowID: subW.ID})
		entryNode := subW.Nodes[subW.Entry]
		e.emit(EventNodeEnter, Event{Type: EventNodeEnter, NodeID: entryNode.ID, WorkflowID: subW.ID})

		return StepResult{Status: StepInvoked, Workflow: subW, Node: entryNode}, nil
	}
	e.skipInvocation = false

	// -- Guard evaluation --
	reader := e.buildBlackboardReader()
	validEdges, err := FilterEdges(e.currentNodeID, w.Edges, reader)
	if err != nil {
		e.status = StatusSuspended
		e.emit(EventEngineError, Event{Type: EventEngineError, NodeID: e.currentNodeID, Reason: err.Error()})
		return StepResult{Status: StepSuspended, Reason: "guard evaluation error"}, nil
	}

	// -- Build DecisionContext and call agent --
	dc := DecisionContext{
		Workflow:   w,
		Node:       node,
		Blackboard: reader,
		ValidEdges: validEdges,
		Stack:      e.stackSnapshot(),
	}

	decision, err := e.agent.Resolve(ctx, dc)
	if err != nil {
		e.status = StatusSuspended
		e.emit(EventEngineError, Event{Type: EventEngineError, NodeID: e.currentNodeID, Reason: err.Error()})
		return StepResult{Status: StepSuspended, Reason: "decision agent error"}, nil
	}

	// -- Handle advance --
	if decision.Type == DecisionAdvance {
		var chosenEdge *Edge
		for i := range validEdges {
			if validEdges[i].ID == decision.Edge {
				chosenEdge = &validEdges[i]
				break
			}
		}
		if chosenEdge == nil {
			e.status = StatusSuspended
			e.emit(EventEngineError, Event{
				Type:   EventEngineError,
				NodeID: e.currentNodeID,
				Reason: fmt.Sprintf("invalid edge '%s'", decision.Edge),
			})
			return StepResult{Status: StepSuspended, Reason: "invalid edge selection"}, nil
		}

		e.emit(EventNodeExit, Event{Type: EventNodeExit, NodeID: e.currentNodeID, WorkflowID: e.currentWorkflowID})
		e.emit(EventEdgeTraverse, Event{Type: EventEdgeTraverse, EdgeID: chosenEdge.ID, WorkflowID: e.currentWorkflowID})

		if len(decision.Writes) > 0 {
			source := BlackboardSource{WorkflowID: e.currentWorkflowID, NodeID: e.currentNodeID, StackDepth: len(e.stack)}
			newEntries := e.currentBlackboard.Append(decision.Writes, source)
			e.emit(EventBlackboardWrite, Event{Type: EventBlackboardWrite, Entries: newEntries, WorkflowID: e.currentWorkflowID})
		}

		e.currentNodeID = chosenEdge.To
		nextNode := w.Nodes[chosenEdge.To]
		e.emit(EventNodeEnter, Event{Type: EventNodeEnter, NodeID: nextNode.ID, WorkflowID: e.currentWorkflowID})

		return StepResult{Status: StepAdvanced, Node: nextNode}, nil
	}

	// -- Handle suspend --
	if decision.Type == DecisionSuspend {
		e.status = StatusSuspended
		e.emit(EventEngineSuspend, Event{Type: EventEngineSuspend, Reason: decision.Reason, NodeID: e.currentNodeID})
		return StepResult{Status: StepSuspended, Reason: decision.Reason}, nil
	}

	// -- Handle complete --
	hasOutgoing := false
	for _, edge := range w.Edges {
		if edge.From == e.currentNodeID {
			hasOutgoing = true
			break
		}
	}
	if hasOutgoing {
		e.status = StatusSuspended
		e.emit(EventEngineError, Event{
			Type:   EventEngineError,
			NodeID: e.currentNodeID,
			Reason: "complete at non-terminal node",
		})
		return StepResult{Status: StepSuspended, Reason: "complete at non-terminal node"}, nil
	}

	if len(decision.Writes) > 0 {
		source := BlackboardSource{WorkflowID: e.currentWorkflowID, NodeID: e.currentNodeID, StackDepth: len(e.stack)}
		newEntries := e.currentBlackboard.Append(decision.Writes, source)
		e.emit(EventBlackboardWrite, Event{Type: EventBlackboardWrite, Entries: newEntries, WorkflowID: e.currentWorkflowID})
	}

	// Root workflow complete
	if len(e.stack) == 0 {
		e.status = StatusCompleted
		e.emit(EventEngineComplete, Event{Type: EventEngineComplete, WorkflowID: e.currentWorkflowID})
		return StepResult{Status: StepCompleted}, nil
	}

	// -- Stack pop: sub-workflow done, return to parent --
	childBB := e.currentBlackboard
	frame := e.stack[0]
	e.stack = e.stack[1:]

	parentBB := NewBlackboard(frame.Blackboard...)
	returnSource := BlackboardSource{
		WorkflowID: frame.WorkflowID,
		NodeID:     frame.CurrentNodeID,
		StackDepth: len(e.stack),
	}

	// Execute returnMap
	childReader := childBB.Reader()
	for _, mapping := range frame.ReturnMap {
		val, ok := childReader.Get(mapping.ChildKey)
		if ok {
			newEntries := parentBB.Append(
				[]BlackboardWrite{{Key: mapping.ParentKey, Value: val}},
				returnSource,
			)
			parentW, _ := e.registry.Get(frame.WorkflowID)
			e.emit(EventBlackboardWrite, Event{
				Type: EventBlackboardWrite, Entries: newEntries,
				WorkflowID: parentW.ID,
			})
		}
	}

	e.currentWorkflowID = frame.WorkflowID
	e.currentNodeID = frame.CurrentNodeID
	e.currentBlackboard = parentBB
	e.skipInvocation = true

	parentW, _ := e.registry.Get(frame.WorkflowID)
	invokingNode := parentW.Nodes[frame.CurrentNodeID]

	e.emit(EventWorkflowPop, Event{Type: EventWorkflowPop, WorkflowID: parentW.ID})
	e.emit(EventNodeEnter, Event{Type: EventNodeEnter, NodeID: invokingNode.ID, WorkflowID: parentW.ID})

	return StepResult{Status: StepPopped, Workflow: parentW, Node: invokingNode}, nil
}

// Run steps until completion, suspension, or context cancellation.
func (e *Engine) Run(ctx context.Context) (StepResult, error) {
	for {
		select {
		case <-ctx.Done():
			return StepResult{}, ctx.Err()
		default:
		}

		result, err := e.Step(ctx)
		if err != nil {
			return result, err
		}
		if result.Status == StepCompleted || result.Status == StepSuspended {
			return result, nil
		}
	}
}

// ---------------------------------------------------------------------------
// State Inspection
// ---------------------------------------------------------------------------

// SessionID returns the current session ID.
func (e *Engine) SessionID() string { return e.sessionID }

// Status returns the engine's current lifecycle state.
func (e *Engine) Status() EngineStatus { return e.status }

// CurrentNode returns the current node, or nil if not initialized.
func (e *Engine) CurrentNode() *Node {
	if e.currentWorkflowID == "" || e.currentNodeID == "" {
		return nil
	}
	w, ok := e.registry.Get(e.currentWorkflowID)
	if !ok {
		return nil
	}
	return w.Nodes[e.currentNodeID]
}

// CurrentWorkflow returns the current workflow, or nil if not initialized.
func (e *Engine) CurrentWorkflow() *Workflow {
	if e.currentWorkflowID == "" {
		return nil
	}
	w, _ := e.registry.Get(e.currentWorkflowID)
	return w
}

// Blackboard returns a BlackboardReader over the current scope chain.
func (e *Engine) Blackboard() BlackboardReader {
	return e.buildBlackboardReader()
}

// Stack returns a snapshot of the call stack.
func (e *Engine) Stack() []StackFrame {
	return e.stackSnapshot()
}

// ValidEdges returns the currently valid outgoing edges.
func (e *Engine) ValidEdges() []Edge {
	w := e.CurrentWorkflow()
	if w == nil || e.currentNodeID == "" {
		return nil
	}
	edges, _ := FilterEdges(e.currentNodeID, w.Edges, e.buildBlackboardReader())
	return edges
}

// ---------------------------------------------------------------------------
// Private helpers
// ---------------------------------------------------------------------------

func (e *Engine) emit(eventType EventType, event Event) {
	event.SessionID = e.sessionID
	for _, h := range e.handlers[eventType] {
		h(event)
	}
}

func (e *Engine) buildBlackboardReader() BlackboardReader {
	if e.currentBlackboard == nil {
		return NewBlackboardReader(nil)
	}
	parentScopes := make([][]BlackboardEntry, len(e.stack))
	for i, frame := range e.stack {
		cp := make([]BlackboardEntry, len(frame.Blackboard))
		copy(cp, frame.Blackboard)
		parentScopes[i] = cp
	}
	return e.currentBlackboard.Reader(parentScopes...)
}

func (e *Engine) stackSnapshot() []StackFrame {
	cp := make([]StackFrame, len(e.stack))
	copy(cp, e.stack)
	return cp
}

func generateUUID() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40 // version 4
	b[8] = (b[8] & 0x3f) | 0x80 // variant 2
	return fmt.Sprintf("%s-%s-%s-%s-%s",
		hex.EncodeToString(b[0:4]),
		hex.EncodeToString(b[4:6]),
		hex.EncodeToString(b[6:8]),
		hex.EncodeToString(b[8:10]),
		hex.EncodeToString(b[10:16]),
	)
}
